From 94f28da8409c6059135e89ac64a0839993124155 Mon Sep 17 00:00:00 2001
From: Andreas Schwab <schwab@linux-m68k.org>
Date: Sat, 30 Jan 2010 10:20:59 +0000
Subject: powerpc: TIF_ABI_PENDING bit removal

From: Andreas Schwab <schwab@linux-m68k.org>

commit 94f28da8409c6059135e89ac64a0839993124155 upstream.

Here are the powerpc bits to remove TIF_ABI_PENDING now that
set_personality() is called at the appropriate place in exec.

Signed-off-by: Andreas Schwab <schwab@linux-m68k.org>
Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 arch/powerpc/include/asm/elf.h         |    8 ++------
 arch/powerpc/include/asm/thread_info.h |    2 --
 arch/powerpc/kernel/process.c          |   12 ------------
 3 files changed, 2 insertions(+), 20 deletions(-)

Backported to Debian's 2.6.26 by dann frazier <dannf@debian.org>

diff -urpN linux-source-2.6.26.orig/arch/powerpc/kernel/process.c linux-source-2.6.26/arch/powerpc/kernel/process.c
--- linux-source-2.6.26.orig/arch/powerpc/kernel/process.c	2008-07-13 15:51:29.000000000 -0600
+++ linux-source-2.6.26/arch/powerpc/kernel/process.c	2010-02-04 17:44:40.000000000 -0700
@@ -501,18 +501,6 @@ void exit_thread(void)
 
 void flush_thread(void)
 {
-#ifdef CONFIG_PPC64
-	struct thread_info *t = current_thread_info();
-
-	if (test_ti_thread_flag(t, TIF_ABI_PENDING)) {
-		clear_ti_thread_flag(t, TIF_ABI_PENDING);
-		if (test_ti_thread_flag(t, TIF_32BIT))
-			clear_ti_thread_flag(t, TIF_32BIT);
-		else
-			set_ti_thread_flag(t, TIF_32BIT);
-	}
-#endif
-
 	discard_lazy_cpu_state();
 
 	if (current->thread.dabr) {
diff -urpN linux-source-2.6.26.orig/include/asm-powerpc/elf.h linux-source-2.6.26/include/asm-powerpc/elf.h
--- linux-source-2.6.26.orig/include/asm-powerpc/elf.h	2008-07-13 15:51:29.000000000 -0600
+++ linux-source-2.6.26/include/asm-powerpc/elf.h	2010-02-04 17:46:53.000000000 -0700
@@ -248,14 +248,10 @@ extern int dump_task_altivec(struct task
 #ifdef __powerpc64__
 # define SET_PERSONALITY(ex, ibcs2)				\
 do {								\
-	unsigned long new_flags = 0;				\
 	if ((ex).e_ident[EI_CLASS] == ELFCLASS32)		\
-		new_flags = _TIF_32BIT;				\
-	if ((current_thread_info()->flags & _TIF_32BIT)		\
-	    != new_flags)					\
-		set_thread_flag(TIF_ABI_PENDING);		\
+		set_thread_flag(TIF_32BIT);			\
 	else							\
-		clear_thread_flag(TIF_ABI_PENDING);		\
+		clear_thread_flag(TIF_32BIT);			\
 	if (personality(current->personality) != PER_LINUX32)	\
 		set_personality(PER_LINUX);			\
 } while (0)
diff -urpN linux-source-2.6.26.orig/include/asm-powerpc/thread_info.h linux-source-2.6.26/include/asm-powerpc/thread_info.h
--- linux-source-2.6.26.orig/include/asm-powerpc/thread_info.h	2008-07-13 15:51:29.000000000 -0600
+++ linux-source-2.6.26/include/asm-powerpc/thread_info.h	2010-02-04 17:47:46.000000000 -0700
@@ -119,7 +119,6 @@ static inline struct thread_info *curren
 #define TIF_RESTORE_SIGMASK	13	/* Restore signal mask in do_signal */
 #define TIF_FREEZE		14	/* Freezing for suspend */
 #define TIF_RUNLATCH		15	/* Is the runlatch enabled? */
-#define TIF_ABI_PENDING		16	/* 32/64 bit switch needed */
 
 /* as above, but as bit values */
 #define _TIF_SYSCALL_TRACE	(1<<TIF_SYSCALL_TRACE)
@@ -137,7 +136,6 @@ static inline struct thread_info *curren
 #define _TIF_RESTORE_SIGMASK	(1<<TIF_RESTORE_SIGMASK)
 #define _TIF_FREEZE		(1<<TIF_FREEZE)
 #define _TIF_RUNLATCH		(1<<TIF_RUNLATCH)
-#define _TIF_ABI_PENDING	(1<<TIF_ABI_PENDING)
 #define _TIF_SYSCALL_T_OR_A	(_TIF_SYSCALL_TRACE|_TIF_SYSCALL_AUDIT|_TIF_SECCOMP)
 
 #define _TIF_USER_WORK_MASK	( _TIF_SIGPENDING | \
